deepspeed/runtime/zero/stage_1_and_2.py

```diff
@@ -390,11 +390,19 @@ def __init__(self,
             # Note that the params in single_partition_of_fp32_groups is cloned and detached
             # from the origin params of the model.
             if not fp16_master_weights_and_gradients:
-                self.single_partition_of_fp32_groups.append(self.parallel_partitioned_bit16_groups[i][partition_id].to(
-                    self.device).clone().float().detach())
+                weights_partition = self.parallel_partitioned_bit16_groups[i][partition_id].to(
```

Comment:
The new change seems unduly complicated with the duplication of the offload condition in the `if not fp16_master_weights_and_gradients` branch. 

Is there a reason why something like below does not work?

```python
            if not fp16_master_weights_and_gradients:
                weights_partition = self.parallel_partitioned_bit16_groups[i][partition_id].to(
                    self.device).clone().float().detach()
            else:
                weights_partition = self.parallel_partitioned_bit16_groups[i][partition_id].to(
                    self.device).clone().half().detach()
           
            if self.cpu_offload:
               weights_partition = get_accelerator().pin_memory(weights_partition)
             
            self.single_partition_of_fp32_groups.append(weights_partition)
``